### PR TITLE
[env] fix: mitigate CPU memory leaks in data serialization over Ray

### PR DESCRIPTION
**Pull request by Sky-Trigger**
# PR title

[examples] fix: enable NumPy serialization for Qwen-Image NPU training

### What does this PR do?

This PR mitigates persistent shard-sized CPU memory growth in
RewardLoopWorkers by setting the following environment variable in the
Qwen-Image NPU training examples:

```bash
export VERL_DATAPROTO_SERIALIZATION_METHOD=numpy
```

This selects verl's existing NumPy-based `DataProto` serialization mode for
large batches transferred through Ray.

The accumulation is observed on the default `DataProto` serialization path in
`verl/protocol.py`. When Ray serializes a `DataProto` argument,
`DataProto.__getstate__()` consolidates its `TensorDict`, writes the complete
batch to an in-memory buffer with `torch.save()`, and returns the resulting
`bytes` payload. In the receiving process, `DataProto.__setstate__()` wraps the
payload in `BytesIO` and reconstructs the tensors with `torch.load()`:

```text
TaskRunner
  DataProto.__getstate__()
    TensorDict -> torch.save() -> BytesIO -> bytes
  -> Ray RPC
RewardLoopWorker
  DataProto.__setstate__()
    bytes -> BytesIO -> torch.load() -> TensorDict
```

In the measured workload, TaskRunner RSS rose to a high-water plateau of about
`57 GiB`. Each persistent RewardLoopWorker, however, continued to gain about
one `928.6 MB` input shard per reward call. The TaskRunner plateau stopped
growing in the observed interval; the Worker growth did not.

Weak-reference probes showed that the tracked reward inputs, tensor wrappers,
chunks, and asynchronous reward tasks were already dead while Worker RSS
remained high. This associates the persistent Worker growth with the default
binary serialization/transport boundary rather than the visual reward
calculation itself. The probes do not identify the exact internal Ray or
PyTorch native owner.

With NumPy serialization, each tensor is transferred as its dtype, shape, and
a NumPy byte view instead of one large `torch.save()` byte stream. The previous
approximately `0.9 GiB` per-call RewardLoopWorker growth was no longer observed
in the tested Qwen-Image FlowGRPO workload. TaskRunner still rose toward its
high-water RSS plateau, so this PR does not claim to eliminate TaskRunner
retention.

This PR changes only the Qwen-Image NPU examples for which the mitigation was
measured. Other workloads and platforms have not yet been tested.

### Checklist Before Starting

- [x] Search for similar PRs.
- [x] Format the PR title as `[{modules}] {type}: {description}`.

### Test

The tests used the same Qwen-Image FlowGRPO workload with a rollout batch of
512 samples. The tensor payload passed into reward computation was
`3714.254 MB` in total and was split across four persistent Ray
RewardLoopWorkers, giving each Worker a `928.564 MB` input shard.

#### NumPy serialization disabled: default serialization

The baseline did not set `VERL_DATAPROTO_SERIALIZATION_METHOD`.

TaskRunner RSS and anonymous RSS after reward computation grew as follows
before reaching a high-water plateau:

| Step | TaskRunner RSS (GiB) | TaskRunner anonymous RSS (GiB) |
| ---: | ---: | ---: |
| 1 | 16.828 | 12.721 |
| 2 | 24.013 | 19.902 |
| 5 | 39.560 | 35.438 |
| 9 | 53.293 | 49.171 |
| 18 | 56.550 | 52.428 |
| 30 | 57.275 | 53.271 |

From steps 1 through 9, TaskRunner RSS increased by `36.465 GiB` and
anonymous RSS increased by `36.450 GiB`. From steps 18 through 30 it remained
near `57 GiB`, showing high-water retention rather than continued linear
growth.

All four receiving RewardLoopWorkers showed persistent growth throughout the
same run:

| Worker PID | Initial RSS (GiB) | Final RSS (GiB) | RSS increase (GiB) | Initial anonymous RSS (GiB) | Final anonymous RSS (GiB) |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 36419 | 4.080 | 29.738 | 25.658 | 2.739 | 28.513 |
| 36420 | 3.919 | 30.029 | 26.110 | 2.577 | 28.804 |
| 36421 | 3.970 | 29.882 | 25.912 | 2.629 | 28.659 |
| 36422 | 5.105 | 29.192 | 24.087 | 3.764 | 27.876 |

The four Workers accumulated approximately `101.8 GiB` of RSS in total. Their
per-call increases were typically about `0.8-1.0 GiB`, closely matching the
`928.564 MB` input shard.

#### NumPy serialization enabled

The comparison run set:

```bash
VERL_DATAPROTO_SERIALIZATION_METHOD=numpy
```

The previous shard-sized Worker growth was not observed. For example, Worker
PID `407093` reported:

| Call | Batch start RSS (kB) | Batch end RSS (kB) | Increase (MiB) |
| ---: | ---: | ---: | ---: |
| 1 | 3,536,600 | 3,838,532 | 294.9 |
| 2 | 3,839,188 | 3,865,768 | 26.0 |

The first call included one-time warm-up growth. The next call increased by
only about `26 MiB`, rather than retaining another approximately `928.6 MB`
input shard. This is substantially different from the baseline's persistent
`0.8-1.0 GiB` increase per call.

TaskRunner continued to rise toward bounded high-water plateaus in this run.
This PR therefore does not claim that NumPy serialization eliminates all
TaskRunner high-water retention; it removes the previous recurring growth of
approximately one complete `3.7 GB` reward batch per step.

#### Extended NumPy run to step 30

A subsequent run with NumPy serialization was observed through step 30. The
memory probes did not show the previous batch-sized growth during this
interval.

After initial warm-up, TaskRunner memory remained in two bounded plateaus:

| Phase | TaskRunner RSS (GiB) | TaskRunner anonymous RSS (GiB) |
| --- | ---: | ---: |
| Initial reward submission | 9.547 | 5.464 |
| First stable plateau | 14.86-14.99 | 10.73-10.86 |
| Later stable plateau | 16.38-16.50 | 12.31-12.38 |
| Last sampled cleanup | 16.422 | 12.380 |

The TaskRunner cleanup probes reported `tensor_mb=0.002` and no recurring
increase of approximately `3.7 GB` per reward call.

The four persistent RewardLoopWorkers also remained near their post-call
baselines through the run:

| Worker PID | First sampled post-call RSS (GiB) | Last sampled post-call RSS (GiB) | First anonymous RSS (GiB) | Last anonymous RSS (GiB) |
| ---: | ---: | ---: | ---: | ---: |
| 35253 | 3.052 | 3.219 | 1.712 | 1.969 |
| 35254 | 2.947 | 3.120 | 1.609 | 1.871 |
| 35255 | 3.146 | 3.198 | 1.809 | 1.948 |
| 35256 | 3.056 | 3.087 | 1.720 | 1.838 |

Every sampled Worker cleanup still reported one large anonymous region
(`large=1/0.47`) and dead tracked inputs (`alive=none`). The small post-call
baseline changes above are not proportional to the `928.564 MB` shard and are
substantially different from the approximately `24-26 GiB` per-Worker
accumulation in the baseline run.

#### Configuration comparison

| Configuration | RewardLoopWorker result | TaskRunner result |
| --- | --- | --- |
| NumPy disabled (default `torch.save` serialization) | Persistent growth of approximately one 928.6 MB shard per call | Rose to a high-water plateau around 57 GiB |
| NumPy enabled | The shard-sized per-call growth was not observed; all four Workers remained near their post-call baselines through step 30 | Settled into bounded plateaus after warm-up; no recurring 3.7 GB per-step growth was observed |

### API and Usage Example

There is no Python API change. The existing NumPy serialization implementation
is selected through:

```bash
VERL_DATAPROTO_SERIALIZATION_METHOD=numpy
```

Leaving the variable unset selects the original `torch.save()` path.

### Design & Code Changes

- Export `VERL_DATAPROTO_SERIALIZATION_METHOD=numpy` from the six Qwen-Image
  NPU example launch scripts.
- Reuse verl's existing NumPy serialization implementation without changing
  `DataProto`, Ray, or reward-loop code.
- Leave the serialization behavior of other examples unchanged.
- Avoid reward-loop-specific deletion or field filtering because downstream
  actor training still needs the complete rollout batch.

### CODE REVIEW

**Comment by zhtmike** [comment]
a regression test should be added to guard this

**Comment by Sky-Trigger** [comment]
Could you clarify what kind of regression test you expect? Detecting this memory leak requires running multiple training steps and monitoring CPU memory growth over time.

**Comment by zhtmike** [comment]
is it possible to write a test with few steps, monitor the mem usage ?

**Review by SamitHuang — APPROVED** [review]
LGTM

**Comment by Sky-Trigger** [comment]
@zhtmike Not easily. So far, the memory growth has only been reproducible during an actual training run. If u want, I can provide the instrumented code with memory probes for reproducing and monitoring the issue.

**Comment by zhtmike** [comment]
> @zhtmike Not easily. So far, the memory growth has only been reproducible during an actual training run. If u want, I can provide the instrumented code with memory probes for reproducing and monitoring the issue.

OK.

**Comment by Sky-Trigger** [comment]
In fact, this issue is not specific to Qwen-Image or NPU; it appears to be a broader Ray issue affecting other workloads as well. However, I’m not sure where this change should be applied globally.

**Comment by zhtmike** [comment]
> In fact, this issue is not specific to Qwen-Image or NPU; it appears to be a broader Ray issue affecting other workloads as well. However, I’m not sure where this change should be applied globally.

can we observe it in wandb log? like `Process Memory Available (MB)`, in my past exp, it stays stable like for 90 steps in GPU environment.
